### PR TITLE
FIX: Insert Decimal with executemany throw an exception

### PR DESCRIPTION
**Pull request by subrata-ms**
### Work Item / Issue Reference  

> [AB#39795](https://sqlclientdrivers.visualstudio.com/c6d89619-62de-46a0-8b46-70b92a84d85e/_workitems/edit/39795)

> GitHub Issue: #272

-------------------------------------------------------------------
### Summary   
  

This pull request improves type inference for column values in the MSSQL Python cursor and adds a new test for handling lists of `Decimal` values in executemany operations. The main changes are focused on enhancing how column types are computed, especially for decimal values, and increasing test coverage for edge cases.

Enhancements to type inference logic:

* Improved the `_compute_column_type` method in `cursor.py` to better handle objects with length (like strings and bytes) and `decimal.Decimal` values, preferring longer objects and higher-precision decimals for sample value selection. This also ensures that decimals are preferred over non-decimals for more accurate type inference.

Test coverage improvements:

* Added a new test `test_executemany_Decimal_list` in `test_004_cursor.py` to verify correct behavior when using `executemany` with lists of `decimal.Decimal` values. This test ensures that the operation handles decimal data types properly and cleans up the test table afterward.

### CODE REVIEW

File: mssql_python/cursor.py [hunk 1]
```diff
@@ -1717,9 +1717,19 @@ def _compute_column_type(self, column):
...
+            if not sample_value:
+                sample_value = v
+            elif hasattr(v, "__len__") and hasattr(sample_value, "__len__") and len(v) > len(sample_value):
+                # For objects with length (strings, bytes, etc.), prefer the longer one
+                sample_value = v
+            elif isinstance(v, decimal.Decimal) and isinstance(sample_value, decimal.Decimal):
+                # For Decimal objects, prefer the one with higher precision (more digits)
+                v_tuple = v.as_tuple()
+                sample_tuple = sample_value.as_tuple()
+                if len(v_tuple.digits) > len(sample_tuple.digits):
```
Comment (sumitmsft) [inline mssql_python/cursor.py]: The current logic for choosing a representative Decimal compares only the count of digits `(len(v.as_tuple().digits))`, ignoring scale (`-exponent` when `exponent < 0`). This can under-estimate required buffer size for parameter binding.

Two values with identical digit counts but different scale produce different string lengths (e.g. Decimal('12345') vs Decimal('123.45')). Binding logic often needs both precision and scale; choosing a lower-scale Decimal may cause truncated allocation and the runtime error you observed.

Please update decimal representative selection to use both precision and scale (from `Decimal.as_tuple()`), not just digit count, otherwise buffer sizing can be underestimated for values like 123.45 vs 12345

File: mssql_python/cursor.py
Comment (sumitmsft) [inline mssql_python/cursor.py:1749]: I was just reviewing this function and found something peculiar. 

The current logic returns immediately if it finds any `int`. In practice, the “column” here represents the batched parameter values for a single placeholder across executemany, not necessarily a single homogeneous SQL column source. It’s common for such a list to mix int and Decimal:

- Whole numbers entered without a fractional part remain int.
- Values with fractional digits (e.g. "2.50") are parsed/converted to Decimal.
- Mixed ingestion from CSV/JSON/Pandas, or business logic that only wraps fractional values in Decimal.

Returning early on integers masks the presence of later Decimal values, causing us to:
1. Lose scale information.
2. Under-size parameter buffers.
3. Potentially hit the observed runtime error (“Input string exceeds allowed column size”).

Please scan entire batch, prioritize decimals when present, and compare (precision, scale).

File: mssql_python/cursor.py [hunk 2]
```diff
@@ -1717,9 +1717,19 @@ def _compute_column_type(self, column):
 
         sample_value = None
         for v in non_nulls:
-            if not sample_value or (
-                hasattr(v, "__len__") and len(v) > len(sample_value)
-            ):
+            if not sample_value:
+                sample_value = v
+            elif hasattr(v, "__len__") and hasattr(sample_value, "__len__") and len(v) > len(sample_value):
```
Comment (sumitmsft) [inline mssql_python/cursor.py]: The function `_compute_column_type` calls `len(v)` after only checking `hasattr(v, "__len__")`. That’s not always safe.

Why this is a problem:

- A custom object’s `__len__` can raise an exception.
- Computing length might be expensive (lazy or remote data).
- Bad implementations may return invalid values (causing `TypeError`).
- One failing `len()` breaks the whole type inference and can lead to wrong buffer sizing later.

Recommended Fixes:
- Wrap the `len()` call in `try/except`; if it fails, treat length as unavailable (e.g. `l = -1`).
- Cache `len(sample_value)` instead of recomputing it repeatedly.

File: tests/test_004_cursor.py [hunk 3]
```diff
@@ -1663,6 +1663,22 @@ def test_executemany_empty_parameter_list(cursor, db_connection):
         db_connection.commit()
 
 
+def test_executemany_Decimal_list(cursor, db_connection):
+    """Test executemany with an empty parameter list."""
```
[the remark this answers is not in the document]
Comment (sumitmsft) [reply tests/test_004_cursor.py]: Pls see if this can be changed @subrata-ms

File: tests/test_004_cursor.py [hunk 4]
```diff
@@ -1663,6 +1663,22 @@ def test_executemany_empty_parameter_list(cursor, db_connection):
...
+    try:
+        cursor.execute("CREATE TABLE #pytest_empty_params (val DECIMAL(30, 20))")
+        data = [(decimal.Decimal('35.1128407822'),), (decimal.Decimal('40000.5640564065406'),)]
+        cursor.executemany("INSERT INTO #pytest_empty_params VALUES (?)", data)
+        db_connection.commit()
+
+        cursor.execute("SELECT COUNT(*) FROM #pytest_empty_params")
+        count = cursor.fetchone()[0]
+        assert count == 0
```
[the remark this answers is not in the document]
Comment (sumitmsft) [reply tests/test_004_cursor.py]: Please validate this as well

File: tests/test_004_cursor.py [hunk 5]
```diff
@@ -1663,6 +1663,22 @@ def test_executemany_empty_parameter_list(cursor, db_connection):
         db_connection.commit()
 
 
+def test_executemany_Decimal_list(cursor, db_connection):
```
Comment (sumitmsft) [inline tests/test_004_cursor.py:1666]: Please see if we can add following test cases:

1. Ensure scale is considered (not just digit count). Data: [Decimal('12345'), Decimal('123.45')] Expect: Representative = Decimal('123.45') (higher scale with same precision).

2. Verify removal of early integer short‑circuit; decimals take precedence. Data: [1, 2, Decimal('3.50')] Expect: Representative is the Decimal; min/max ints are not returned.

3. Stress buffer sizing at upper bound. Data: [Decimal('9' * 38), Decimal('1.' + '2' * 37)] (e.g. precision 38, scale 37) Expect: Representative = second value if comparison is (precision, scale); no overflow/truncation in binding.

4. Ensure scale counts fractional zeros (impacts length). Data: [Decimal('123.4500'), Decimal('123.45'), Decimal('123.4560')] Expect: Representative = Decimal('123.4560') (highest (precision, scale) or longest effective fractional part).

5. Verify that a batch with varying scales binds without error. SQL Column: DECIMAL(10,4) Batch: [Decimal('1.2345'), Decimal('9999.0000')] Expect: executemany succeeds; rowcount = 2.

File: tests/test_004_cursor.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (subrata-ms) [reply tests/test_004_cursor.py]: part of next commit.

File: tests/test_004_cursor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (subrata-ms) [reply tests/test_004_cursor.py]: Part of next commit

**Comment by subrata-ms** [comment]
@microsoft-github-policy-service agree company="Microsoft"

File: mssql_python/cursor.py (hunk 1 above)
Comment (subrata-ms) [reply mssql_python/cursor.py]: Thanks for pointing this. Fixed in next commit.

File: mssql_python/cursor.py (hunk 2 above)
Comment (subrata-ms) [reply mssql_python/cursor.py]: Made the changes. Should be available in next commit.

File: tests/test_004_cursor.py (hunk 5 above)
Comment (subrata-ms) [reply tests/test_004_cursor.py:1666]: Thanks for the detailed review. I have added test cases for 1,4, 5 and additional string of decimal test cases.

File: mssql_python/cursor.py
Comment (subrata-ms) [reply mssql_python/cursor.py:1749]: As discusses, i will create a separate github bug for this scenario. This is not working currently and throwing an exception in ddbc layer.
